### DISCUSSION TITLE
Creating an Archetype through FFI?

### DISCUSSION THREAD

**Discussion by safasofuoglu** [bevyengine/bevy#1271] [Q&A] [answered]
Hello! I am experimenting with building a scripting layer on top of bevy (another language, through FFI). I know what I'm describing is a non-goal for bevy, just looking to figure out whether I can build on top of bevy or write a custom ECS.

Problem: 
- On the scripting side, I am able to create C-style structs but not bevy components or archetypes, since archetypes need TypeId, TypeInfo, TypeState etc.

Current approach:
- Build a rust compatibility layer between bevy and scripting, exposing a C api
- Provide TypeId, TypeInfo, TypeState on the compatibility layer and fake them where necessary. For example, create dummy rust placeholder types (Type1, Type2) and map their TypeId's to structs on the scripting side. Not sure if I can meet all requirements though.

@cart do you think it's theoretically possible to create archetypes this way? If not, would you suggest a way to fork/modify bevy_ecs to make it work?

**Comment by cart** [comment, marked as answer]
You might be interested #623 and #635. I'm also currently working on a Bevy ECS core rewrite, which should make this type of integration even cleaner.
